test/async-hooks/test-async-local.js

```diff
@@ -0,0 +1,173 @@
+    assert.strictEqual(asyncLocal1.value, 'one');
+    assert.strictEqual(asyncLocal2.value, 'two');
+    assert.strictEqual(asyncLocal3.value, 'three');
+    setImmediate(common.mustCall(() => {
+      assert.strictEqual(asyncLocal1.value, 'one');
+      assert.strictEqual(asyncLocal2.value, 'two');
+      assert.strictEqual(asyncLocal3.value, 'three');
+    }));
+    asyncLocal1.value = null;
+    asyncLocal3.value = 'four';
+    assert.strictEqual(asyncLocal1.value, undefined);
+    assert.strictEqual(asyncLocal2.value, 'two');
+    assert.strictEqual(asyncLocal3.value, 'four');
+    setImmediate(common.mustCall(() => {
+      assert.strictEqual(asyncLocal1.value, undefined);
+      assert.strictEqual(asyncLocal2.value, 'two');
+      assert.strictEqual(asyncLocal3.value, 'four');
+    }));
+  }));
+}
+
+{
+  async function asyncFunc() {}
+
+  const asyncLocal = new AsyncLocal();
+
+  async function testAwait() {
+    asyncLocal.value = 42;
+    await asyncFunc();
+    assert.strictEqual(asyncLocal.value, 42);
+  }
+  testAwait().then(common.mustCall(() =>
+    assert.strictEqual(asyncLocal.value, 42)
+  ));
+}
+
+{
+  const asyncLocal = new AsyncLocal();
+  const mutableObj = { a: 'b' };
+
+  asyncLocal.value = mutableObj;
+  process.nextTick(common.mustCall(() => {
+    assert.deepStrictEqual(mutableObj, { a: 'b', b: 'a' });
+  }));
+  mutableObj.b = 'a';
+}
+
+{
+  const exp = [
+    [ undefined, 'foo', false ],
+    [ 'foo', undefined, false ],
+    [ undefined, 'bar', false ]
+  ];
+
+  const act = [];
+  const asyncLocal = new AsyncLocal({
+    onChangedCb: (p, c, t) => act.push([p, c, t])
+  });
+
+  asyncLocal.value = 'foo';
+  assert.strictEqual(act.length, 1);
+
+  asyncLocal.value = null;
+  assert.strictEqual(act.length, 2);
+
+  asyncLocal.value = 'bar';
+  assert.strictEqual(act.length, 3);
+
+  assert.deepStrictEqual(act, exp);
+}
+
+{
+  const asyncLocal = new AsyncLocal();
+  const asyncRes1 = new AsyncResource('Resource1');
+  asyncLocal.value = 'R';
+  const asyncRes2 = new AsyncResource('Resource2');
+
+  asyncRes1.runInAsyncScope(common.mustCall(() => {
+    assert.strictEqual(asyncLocal.value, undefined);
+    asyncRes2.runInAsyncScope(common.mustCall(() =>
+      assert.strictEqual(asyncLocal.value, 'R')
+    ));
+    assert.strictEqual(asyncLocal.value, undefined);
+  }));
+  assert.strictEqual(asyncLocal.value, 'R');
+}
+
+{
+  const exp = [
+    [ undefined, 'foo', false ],
+    [ 'foo', 'bar', false ],
+    [ 'bar', 'foo', true ],
+    [ 'foo', 'bar', true ],
+    [ 'bar', 'foo', true ],
+    [ 'foo', undefined, true ],
+    [ undefined, 'foo', true ],
+    [ 'foo', undefined, true ],
+    [ undefined, 'bar', true ],
+  ];
+
+  const act = [];
+  const asyncLocal = new AsyncLocal({
+    onChangedCb: (p, c, t) => act.push([p, c, t])
+  });
+
+  process.nextTick(common.mustCall(() => {
+    asyncLocal.value = 'foo';
+    assert.strictEqual(act.length, 1);
+
+    const r1 = new AsyncResource('R1');
```

Comment:
It might be worthwhile to also test that behavior is correct if we enter the same AsyncResource multiple times, which AFAICT is not represented in this test suite (correct me if I'm wrong).

Reply:
added a test which verifies entering same resource again; directly and with another one in between.